Risheek Chemudupati
Data Visualization

# Milestone Project: Data Visualization - Prototype Code, Documentation, Preliminary Findings

# Importing Packages

In [26]:
import numpy as np
import seaborn as sns
import pandas as pd
import matplotlib.pyplot as plt
import plotly as plx
import altair as alt

# Input and Modifying the Data

Here, I modified the data to seperate the systolic and diastolic blood pressure. In addition, I did some data cleaning by replacing null values in Sleep Disorder and BMI Category with "None" or uniformly making "Normal Weight" into "Normal."

The dataset comes from: https://data.mendeley.com/datasets/46j8wrc7p7/1. On Google Colab, you must upload it in the file repository but you may modify the file path for your needs.

In [27]:
filepath = "/content/Sleep_health_and_lifestyle_dataset.csv"

In [28]:
df = pd.read_csv(filepath)
df['Sleep Disorder'] = (df['Sleep Disorder'].replace(np.nan, 'None'))
df['BMI Category'] = (df['BMI Category'].replace('Normal Weight', 'Normal'))
bloodPressure = df['Blood Pressure'].str.split(pat='/', expand=True)
df['Systolic Blood Pressure'] = bloodPressure[0]
df['Diastolic Blood Pressure'] = bloodPressure[1]
df.head()

Person ID Gender  Age            Occupation  Sleep Duration  \
0          1   Male   27     Software Engineer             6.1   
1          2   Male   28                Doctor             6.2   
2          3   Male   28                Doctor             6.2   
3          4   Male   28  Sales Representative             5.9   
4          5   Male   28  Sales Representative             5.9   

   Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                 6                       42             6   Overweight   
1                 6                       60             8       Normal   
2                 6                       60             8       Normal   
3                 4                       30             8        Obese   
4                 4                       30             8        Obese   

  Blood Pressure  Heart Rate  Daily Steps Sleep Disorder  \
0         126/83          77         4200           None   
1         125/80          75        10000           None   
2         125/80          75        10000           None   
3         140/90          85         3000    Sleep Apnea   
4         140/90          85         3000    Sleep Apnea   

  Systolic Blood Pressure Diastolic Blood Pressure  
0                     126                       83  
1                     125                       80  
2                     125                       80  
3                     140                       90  
4                     140                       90

# Basic Descriptive Analyses

First, I looked at the over all summary table using the describe function. This is not useful for the categorical variables in some instances but it helps to see an overhead view of all the data.

In [29]:
df.describe()

Person ID         Age  Sleep Duration  Quality of Sleep  \
count  374.000000  374.000000      374.000000        374.000000   
mean   187.500000   42.184492        7.132086          7.312834   
std    108.108742    8.673133        0.795657          1.196956   
min      1.000000   27.000000        5.800000          4.000000   
25%     94.250000   35.250000        6.400000          6.000000   
50%    187.500000   43.000000        7.200000          7.000000   
75%    280.750000   50.000000        7.800000          8.000000   
max    374.000000   59.000000        8.500000          9.000000   

       Physical Activity Level  Stress Level  Heart Rate   Daily Steps  
count               374.000000    374.000000  374.000000    374.000000  
mean                 59.171123      5.385027   70.165775   6816.844920  
std                  20.830804      1.774526    4.135676   1617.915679  
min                  30.000000      3.000000   65.000000   3000.000000  
25%                  45.000000      4.000000   68.000000   5600.000000  
50%                  60.000000      5.000000   70.000000   7000.000000  
75%                  75.000000      7.000000   72.000000   8000.000000  
max                  90.000000      8.000000   86.000000  10000.000000

Then, I looked at the descriptive analysis for the continuous variables.

In [30]:
df['Age'].describe()

count    374.000000
mean      42.184492
std        8.673133
min       27.000000
25%       35.250000
50%       43.000000
75%       50.000000
max       59.000000
Name: Age, dtype: float64

In [31]:
df['Physical Activity Level'].describe()


count    374.000000
mean      59.171123
std       20.830804
min       30.000000
25%       45.000000
50%       60.000000
75%       75.000000
max       90.000000
Name: Physical Activity Level, dtype: float64

In [32]:
df['Sleep Duration'].describe()


count    374.000000
mean       7.132086
std        0.795657
min        5.800000
25%        6.400000
50%        7.200000
75%        7.800000
max        8.500000
Name: Sleep Duration, dtype: float64

In [33]:
df['Heart Rate'].describe()

count    374.000000
mean      70.165775
std        4.135676
min       65.000000
25%       68.000000
50%       70.000000
75%       72.000000
max       86.000000
Name: Heart Rate, dtype: float64

In [34]:
df['Daily Steps'].describe()

count      374.000000
mean      6816.844920
std       1617.915679
min       3000.000000
25%       5600.000000
50%       7000.000000
75%       8000.000000
max      10000.000000
Name: Daily Steps, dtype: float64

Then I looked at the count for each of the categorical variables.

In [37]:
df['BMI Category'].value_counts()

BMI Category
Normal        216
Overweight    148
Obese          10
Name: count, dtype: int64

In [38]:
df['Stress Level'].value_counts()

Stress Level
3    71
8    70
4    70
5    67
7    50
6    46
Name: count, dtype: int64

In [39]:
df['Gender'].value_counts()

Gender
Male      189
Female    185
Name: count, dtype: int64

In [40]:
df['Occupation'].value_counts()

Occupation
Nurse                   73
Doctor                  71
Engineer                63
Lawyer                  47
Teacher                 40
Accountant              37
Salesperson             32
Scientist                4
Software Engineer        4
Sales Representative     2
Manager                  1
Name: count, dtype: int64

In [41]:
df['Sleep Disorder'].value_counts()

Sleep Disorder
None           219
Sleep Apnea     78
Insomnia        77
Name: count, dtype: int64

# Prototype Code

## Prototype Code

In [ ]:
# We will have Sleep Duration avs. Sleep Quality Blood Pressure, BMI, Heart Rate, Physical Activity Level - able to be sperated by Gender, Occupation, and Sleep Disorder as options

y_var = 'Quality of Sleep:Q'
x_var = 'Sleep Duration:Q'
y = 'Quality of Sleep'
x = 'Sleep Duration'

sq_width = 500
sq_height = 500
tip = ['Gender:N','Age:Q', 'Occupation:N','Sleep Duration:Q', 'Daily Steps:Q', 'Physical Activity Level:Q', 'BMI Category:N','Stress Level:Q']

sleepDisorder_select = alt.selection_point(fields=['Sleep Disorder'])
color_sleepDisorder= (alt.when(sleepDisorder_select).then(alt.value("lightgreen")).otherwise(alt.value("red")))

legend = alt.Chart(df).mark_rect().encode(
    alt.X('Sleep Disorder'),
    color = color_sleepDisorder,
    tooltip=alt.Tooltip(['count()'])
).add_params(sleepDisorder_select)

dropdown_vs_occ_select = alt.selection_point(fields=['Occupation'])
color_occ= (alt.when(dropdown_vs_occ_select).then(alt.value("lightgreen")).otherwise(alt.value("red")))

legend2 = alt.Chart(df).mark_rect().encode(
    alt.X('Occupation'),
    color = color_occ,
    tooltip=alt.Tooltip(['count()'])
).add_params(dropdown_vs_occ_select)


bmi_select = alt.selection_point(fields=['BMI Category'])
color_bmi= (alt.when(bmi_select).then(alt.value("lightgreen")).otherwise(alt.value("red")))

legend3 = alt.Chart(df).mark_rect().encode(
    alt.X('BMI Category:N'),
    color = color_bmi,
    tooltip=alt.Tooltip(['count()'])
).add_params(
    bmi_select
)

gender_select = alt.selection_point(fields=['Gender'])
color_gender= (alt.when(gender_select).then(alt.value("lightgreen")).otherwise(alt.value("red")))

legend4 = alt.Chart(df).mark_rect().encode(
    alt.X('Gender:N'),
    color = color_gender,
    tooltip=alt.Tooltip(['count()'])
).add_params(
    gender_select
)

points = alt.Chart(df).encode(
    alt.X(x_var),
    alt.Y(y_var),
    tooltip=alt.Tooltip(tip)
).properties(width=sq_width, height=sq_height).interactive().add_params(
    dropdown_vs_occ_select,bmi_select,sleepDisorder_select, gender_select
).transform_filter(
    dropdown_vs_occ_select,bmi_select,sleepDisorder_select, gender_select
)

all_points = alt.Chart(df).add_params(
    dropdown_vs_occ_select,bmi_select,sleepDisorder_select,gender_select
).transform_filter(
    dropdown_vs_occ_select,bmi_select,sleepDisorder_select,gender_select
)

chart_bp_a = all_points.encode(y='mean(Systolic Blood Pressure):Q', x='Quality of Sleep:Q').properties(width=300, height=300).mark_bar(color='red', width=35)
chart_bp_b = all_points.encode(y='mean(Diastolic Blood Pressure):Q', x='Quality of Sleep:Q').properties(width=300, height=300).mark_bar(color='blue', width=35)
chart_bp = chart_bp_a | chart_bp_b

chart2 = all_points.encode(y='mean(Physical Activity Level):Q', x='Quality of Sleep:Q').properties(width=100, height=100).mark_bar(color='black', width=15)
chart3 = all_points.encode(y='mean(Stress Level):Q', x='Quality of Sleep:Q').properties(width=100, height=100).mark_bar(color='black', width=15)
chart4 = all_points.encode(y='mean(Heart Rate):Q', x='Quality of Sleep:Q').properties(width=100, height=100).mark_bar(color='black', width=15)
chart5 = all_points.encode(y='mean(Daily Steps):Q', x='Quality of Sleep:Q').properties(width=100, height=100).mark_bar(color='black', width=15)
chart6 = all_points.encode(y='mean(Sleep Duration):Q', x='Quality of Sleep:Q').properties(width=100, height=100).mark_bar(color='black', width=15)

layer1 =  (chart3 | chart4 | chart5 | chart6 | chart2)
layer1b = (chart_bp)
heatmap1 = alt.Chart(df).encode(
    alt.X('BMI Category'),
    alt.Y('Occupation'),
    color = alt.Color('count()').scale(scheme='purples').legend(None),
    tooltip=alt.Tooltip(['BMI Category','Occupation','count()'])
).properties(width=50, height=100).mark_rect()


heatmap2 = alt.Chart(df).encode(
    alt.X('Sleep Disorder'),
    alt.Y('Occupation'),
    color = alt.Color('count()').scale(scheme='purples').legend(None),
    tooltip=alt.Tooltip(['Sleep Disorder','Occupation','count()'])
).properties(width=50, height=100).mark_rect()


heatmap3 = alt.Chart(df).encode(
    alt.X('Sleep Disorder'),
    alt.Y('BMI Category'),
    color = alt.Color('count():Q').scale(scheme='purples').legend(None),
    tooltip=alt.Tooltip(['Sleep Disorder','BMI Category','count()'])
).properties(width=50, height=100).mark_rect()

heatmap4 = alt.Chart(df).encode(
    alt.X('Gender'),
    alt.Y('Sleep Disorder'),
    color = alt.Color('count():Q').scale(scheme='purples').legend(None),
    tooltip=alt.Tooltip(['Gender','Sleep Disorder','count()'])
).properties(width=50, height=100).mark_rect()

heatmap5 = alt.Chart(df).encode(
    alt.X('Gender'),
    alt.Y('Occupation'),
    color = alt.Color('count():Q').scale(scheme='purples').legend(None),
    tooltip=alt.Tooltip(['Gender','Occupation','count()'])
).properties(width=50, height=100).mark_rect()

layer2 = heatmap1 | heatmap2 | heatmap5| heatmap3 | heatmap4
layer3 = ((legend | legend2 | legend3 | legend4))
allLayers = alt.vconcat(layer1, layer1b, layer3,layer2, center = True).properties(title=alt.TitleParams(text='2024 Health and Sleep Relation Data from University of Oxford', subtitle='Dataset provided by Ben Cutting. Visualization by Risheek Chemudupati.')).configure(padding={'left': 50,'right': 50, 'top':20,'bottom':20})
allLayers.configure(autosize=alt.AutoSizeParams(resize=False))
allLayers.configure_title(fontSize=20, anchor="middle")


# Visualization

This is the code to display the whole visualization.

In [48]:
layer2 = heatmap1 | heatmap2 | heatmap5| heatmap3 | heatmap4
layer3 = ((legend | legend2 | legend3 | legend4))
allLayers = alt.vconcat(layer1, layer1b, layer3,layer2, center = True).properties(title=alt.TitleParams(text='2024 Health and Sleep Relation Data from University of Oxford', subtitle='Dataset provided by Ben Cutting. Visualization by Risheek Chemudupati.')).configure(padding={'left': 50,'right': 50, 'top':20,'bottom':20})
allLayers.configure(autosize=alt.AutoSizeParams(resize=False))
allLayers.configure_title(fontSize=20, anchor="middle")

alt.VConcatChart(...)

# Preliminary Report

## About The Dataset

This was based on a sleep disorder study funded by the University of Oxford to study sleep patterns with multiple data-types concerning age, heart rate, blood pressure, stress level, occupation, and more potentially interesting variables to compare, study, and explore. It was provided by Mendeley Data from Ben Cutting and published on March 20th, 2025.

## Data Cleaning

The total number of participants was 374. Some data-cleaning and processing was done to modify ‘Normal Weight’ into ‘Normal’ as well as null values in the Sleep Disorder category to ‘None’ if it was null. No other values were imputed or replaced to not introduce skew to the results for the analysis, and outliers were not modified for the same reason.

## Descriptive Analysis

The average of the participants was 42.1 + 8.6 years old with a minimum age of 27 and a maximum age of 59. It seems that the average sleep duration was 7.1 hours.  So this would suggest that most participants were roughly middle-aged. The physical activity level was on average 59.17. There was a roughly even gender split, about 189 males and 185 females.  The occupation of the individuals varied with only one manager but also 73 nurses, 71 doctors, and 63 engineers. Most had no sleep disorder (count = 219), and less than 200 individuals had either sleep apnea (count = 78) or insomnia (count = 77).  

## Visualization Analysis

The exploratory visualization revealed that 61 Nurses had sleep apnea, given that 73 nurses were female. There were no males in the following occupations of scientist, nurse, and manager. There were no females in the occupations of sales representative, salesperson, or software engineer. Furthermore, there were 200 individuals who had normal BMI and no sleep disorder. There were 137 males with no sleep disorder compared to 82 females with no sleep disorder. None of the obese individuals had no sleep disorder and this was seen to a lesser degree with those who were overweight. Among those with no sleep disorder, regardless of occupation, BMI, and gender – they tended to have slightly decreasing mean heart rate or increasing mean sleep duration with increasing quality of sleep. Among those with sleep apnea, there was a large spike in mean physical activity, mean daily steps.  This was the same if they were female, were a nurse, or were overweight. Among males, in general, mean heart rate increased if they rated their quality of sleep a 9 compared to the overall trend of decreasing with increased quality of sleep. There was also a drop among this group for mean daily steps at a quality of sleep level of 9. This drop was still also noticed for those with sleep apnea.

# References
https://altair-viz.github.io/index.html

https://pandas.pydata.org/pandas-docs/stable/index.html#

https://data.mendeley.com/datasets/46j8wrc7p7/1

Cutting, Ben (2025), “Health and Sleep relation 2024”, Mendeley Data, V1, doi: 10.17632/46j8wrc7p7.1  

VanderPlas et al., (2018). Altair: Interactive Statistical Visualizations for Python. Journal of Open Source Software, 3(32), 1057, https://doi.org/10.21105/joss.01057  

NumFOCUS, Inc. (2025). pandas. Retrieved from pandas - Python Data Analysis Library: https://pandas.pydata.org/